witwidget/notebook/jupyter/wit.py

```diff
@@ -76,23 +92,163 @@ def render(self):
     return self
 
   def set_examples(self, examples):
+    if self.transfer_block:
+      print('Cannot set examples while transfer is in progress.')
+      return
     base.WitWidgetBase.set_examples(self, examples)
+    self.examples_generator = self.generate_next_example_batch()
+    # If this is called after frontend is ready this makes sure examples are
+    # updated.
+    self._start_examples_sync()
     self._generate_sprite()
 
+  def generate_next_example_batch(self):
+    n_examples = len(self.examples)
+    n_batches = n_examples // self.batch_size
+    batch_end = n_batches * self.batch_size
+    num_remaining = n_batches + (batch_end!=n_examples)
+    for i in range(n_batches):
+      num_remaining -= 1
+      yield self.examples[i*self.batch_size:(i+1)*self.batch_size], num_remaining
+    if batch_end != n_examples:
+      num_remaining -= 1
+      yield self.examples[batch_end:], num_remaining
+
   def _report_error(self, err):
     self.error = {
       'msg': repr(err),
       'counter': self.error_counter
     }
     self.error_counter += 1
 
+  def _start_examples_sync(self):
+    if not self.frontend_ready or self.examples_generator is None or self.transfer_block:
+      return
+    # Send the first batch
+    next_batch, self.examples_batch_id = next(self.examples_generator, ([], -1))
+    self.transfer_block = True
+    self.examples_batch = next_batch
+
   @observe('infer')
   def _infer(self, change):
     try:
       self.inferences = base.WitWidgetBase.infer_impl(self)
+      self.inferences_generator = self.generate_next_inference_batch()
+      self._start_inferences_sync()
     except Exception as e:
       self._report_error(e)
 
+  def generate_next_inference_batch(self):
+    # Parse out the inferences from the returned stucture and empty the
+    # structure of contents, keeping its nested structure.
+    # Chunks of the inference results will be sent to the front-end and
+    # re-assembled.
+    indices = self.inferences['inferences']['indices'][:]
+    self.inferences['inferences']['indices'] = []
+    res2 = []
+    extra = {}
+    extra2 = {}
+    model_inference = self.inferences['inferences']['results'][0]
+    if ('extra_outputs' in self.inferences and len(self.inferences['extra_outputs']) and
+        self.inferences['extra_outputs'][0]):
+      for key in self.inferences['extra_outputs'][0]:
+        extra[key] = self.inferences['extra_outputs'][0][key][:]
+        self.inferences['extra_outputs'][0][key] = []
+    if 'classificationResult' in model_inference:
+      res = model_inference['classificationResult']['classifications'][:]
+      model_inference['classificationResult']['classifications'] = []
+    else:
+      res = model_inference['regressionResult']['regressions'][:]
+      model_inference['regressionResult']['regressions'] = []
+
+    if len(self.inferences['inferences']['results']) > 1:
```

Comment:
nit: this whole structure seems to be repeating for model 1 and 2, you can simplify this with a function if you like.

Reply:
ack